pkg/volume/scaleio/sio_volume.go

```diff
@@ -0,0 +1,407 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scaleio
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/golang/glog"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kubernetes/pkg/api/resource"
+	api "k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/mount"
+	kstrings "k8s.io/kubernetes/pkg/util/strings"
+	"k8s.io/kubernetes/pkg/util/uuid"
+	"k8s.io/kubernetes/pkg/volume"
+	"k8s.io/kubernetes/pkg/volume/util"
+)
+
+type sioVolume struct {
+	sioMgr     *sioMgr
+	plugin     *sioPlugin
+	pod        *api.Pod
+	podUID     types.UID
+	spec       *volume.Spec
+	source     *api.ScaleIOVolumeSource
+	volName    string
+	readOnly   bool
+	options    volume.VolumeOptions
+	configData map[string]string
+
+	volume.MetricsNil
+}
+
+// *******************
+// volume.Volume Impl
+var _ volume.Volume = &sioVolume{}
+
+// GetPath returns the volume directory for the pod
+func (v *sioVolume) GetPath() string {
+	return v.plugin.host.GetPodVolumeDir(
+		v.podUID,
+		kstrings.EscapeQualifiedNameForDisk(sioPluginName),
+		v.volName)
+}
+
+// *************
+// Mounter Impl
+// *************
+var _ volume.Mounter = &sioVolume{}
+
+// CanMount checks to verify that the volume can be mounted prior to Setup.
+// A nil error indicates that the volume is ready for mounitnig.
+func (v *sioVolume) CanMount() error {
+	return nil
+}
+
+func (v *sioVolume) SetUp(fsGroup *int64) error {
+	return v.SetUpAt(v.GetPath(), fsGroup)
+}
+
+// SetUp bind mounts the disk global mount to the volume path.
+func (v *sioVolume) SetUpAt(dir string, fsGroup *int64) error {
+	v.plugin.volumeMtx.LockKey(v.source.VolumeName)
+	defer v.plugin.volumeMtx.UnlockKey(v.source.VolumeName)
+
+	glog.V(4).Info(log("setting up volume %s", v.source.VolumeName))
+	if err := v.setSioMgr(); err != nil {
+		glog.Error(log("setup failed to create scalio manager: %v", err))
+		return err
+	}
+
+	notDevMnt, err := v.plugin.mounter.IsLikelyNotMountPoint(dir)
+	if err != nil && !os.IsNotExist(err) {
+		glog.Error(log("IsLikelyNotMountPoint test failed for dir %v", dir))
+		return err
+	}
+	if !notDevMnt {
+		glog.V(4).Info(log("skipping setup, dir %s already a mount point", v.source.VolumeName))
+		return nil
+	}
+
+	// attach the volume and mount
+	devicePath, err := v.sioMgr.AttachVolume(v.source.VolumeName)
```

Comment:
I started investigated this, on a different branch, to use the `attach/detach` API.  If time permits, I will look into using it.